lib/puppet/context.rb

```diff
@@ -19,66 +21,42 @@ class DuplicateRollbackMarkError < Puppet::Error; end
 
   # @api private
   def initialize(initial_bindings)
-    @table = initial_bindings
-    @ignores = []
-    @description = "root"
-    @id = 0
-    @rollbacks = {}
-    @stack = [[0, nil, nil]]
+    @stack = Concurrent::ThreadLocalVar.new(Puppet::Context::EmptyStack.new.push(initial_bindings))
+    @rollbacks = Concurrent::ThreadLocalVar.new({})
   end
 
   # @api private
   def push(overrides, description = "")
```

Comment:
it feels weird, but description seems to only exist as a comment when this is called, it's not used later

Reply:
It was put in as a debugging aid for manual inspection in a debugger to help figuring out "where is value/context coming from".
While it could be removed, the one time when you really need it, it will be difficult to get the same some other way.
Could be replaced with automatically getting file/line for the caller as the actual description text isn't all that helpful, and armed with file/line you would find comments etc as to the purpose of the push.

Reply:
Okay, I'll make sure the descriptions still get saved. Unfortunately any stack inspection has proved to be extremely expensive on jruby.